File: Back-End/models/receipt/ReceiptBookTransfer.js
// models/receipt/receiptBookTransfer.js
const { nanoid } = require('nanoid');

module.exports = (sequelize, DataTypes) => {
    return sequelize.define('ReceiptBookTransfer', {
        transferID: {
            type: DataTypes.STRING,
            primaryKey: true,
            defaultValue: () => `trans_${nanoid()}`,
        },
        bookID: {
            type: DataTypes.STRING,
            allowNull: false,
            references: { model: 'ReceiptBooks', key: 'bookID' },
        },
        fromUserID: {
            type: DataTypes.STRING,
            allowNull: true, // Null if from agent or initial stock
            references: { model: 'Users', key: 'userID' },
        },
        toUserID: {
            type: DataTypes.STRING,
            allowNull: true, // Null if to agent
            references: { model: 'Users', key: 'userID' },
        },
        toAgentID: {
            type: DataTypes.STRING,
            allowNull: true, // Null if to user
            references: { model: 'Agents', key: 'agentID' },
        },
        status: {
            type: DataTypes.ENUM('Pending', 'Validated'),
            allowNull: false,
            defaultValue: 'Pending',
        },
        transferType: {
            type: DataTypes.ENUM(
                'ToSupplier', 'ToRegionalManager', 'ToSupervisor', 
                'ToAgent', 'StubToSupervisor', 'ToRegionalManagerFromSupervisor', 
                'ToStockManager', 'Archived'
            ),
            allowNull: false,
        },
        transferDate: {
            type: DataTypes.DATE,
            allowNull: false,
            defaultValue: DataTypes.NOW,
        },
    });
};

File: Code/Mobile/lib/models/ReceiptBookTransfer.dart
class ReceiptBookTransfer {
  final String transferID;
  final String bookID;
  final String? fromUserID;
  final String? toUserID;
  final String? toAgentID;
  final String status;
  final String transferType;
  final DateTime transferDate;

  ReceiptBookTransfer({
    required this.transferID,
    required this.bookID,
    this.fromUserID,
    this.toUserID,
    this.toAgentID,
    required this.status,
    required this.transferType,
    required this.transferDate,
  });

  factory ReceiptBookTransfer.fromJson(Map<String, dynamic> json) {
    return ReceiptBookTransfer(
      transferID: json['transferID'] as String,
      bookID: json['bookID'] as String,
      fromUserID: json['fromUserID'] as String?,
      toUserID: json['toUserID'] as String?,
      toAgentID: json['toAgentID'] as String?,
      status: json['status'] as String,
      transferType: json['transferType'] as String,
      transferDate: DateTime.parse(json['transferDate'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'transferID': transferID,
      'bookID': bookID,
      'fromUserID': fromUserID,
      'toUserID': toUserID,
      'toAgentID': toAgentID,
      'status': status,
      'transferType': transferType,
      'transferDate': transferDate.toIso8601String(),
    };
  }
}